<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started with tiledbsc • tiledbsc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with tiledbsc">
<meta property="og:description" content="tiledbsc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tiledbsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.5.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/filtering.html">Filtering</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/quickstart.html">Getting started with tiledbsc</a>
    </li>
    <li>
      <a href="../articles/updates.html">Selective updates</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TileDB-Inc/tiledbsc/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started with tiledbsc</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/TileDB-Inc/tiledbsc/blob/HEAD/vignettes/quickstart.Rmd" class="external-link"><code>vignettes/quickstart.Rmd</code></a></small>
      <div class="hidden name"><code>quickstart.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This vignette provides an overview of the tiledbsc R package which
uses the open source <a href="https://github.com/single-cell-data/SOMA" class="external-link">SOMA data model</a> to
store single-cell data in a collection of TileDB arrays.</p>
<p>With tiledbsc you can:</p>
<ul>
<li>convert data from Seurat and other popular single-cell formats into
SOMA</li>
<li>access SOMA data using any of TileDB’s APIs</li>
<li>store your datasets on S3 (or other remote object stores) and slice
directly without having to download anything first</li>
</ul>
<p>In this vignette you’ll learn how to:</p>
<ul>
<li>ingest Seurat data into TileDB-backed SOMA Collections</li>
<li>access various components of a dataset</li>
<li>slice by cell-/feature-identifiers</li>
<li>query using cell-/feature-level annotations</li>
<li>update a SOMA with analysis results</li>
<li>load data from SOMA into R as a Seurat or Bioconductor (i.e.,
<code>SummarizedExperiment</code>) object</li>
<li>store your data on S3 and register with TileDB Cloud</li>
</ul>
</div>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TileDB-Inc/tiledbsc" class="external-link">tiledbsc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mojaveazure.github.io/seurat-object/" class="external-link">SeuratObject</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span></code></pre></div>
<p>For this tutorial we’ll use a dataset from 10X genomics containing
2,700 peripheral blood mononuclear cells (PBMC) from a healthy donor.
The dataset is available on <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k" class="external-link">10X
Genomics’ website</a>.</p>
<p>Use the provided helper function to download the filtered gene/cell
matrix from 10X and create a <code>Seurat</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataset_seurat_pbmc3k.html">dataset_seurat_pbmc3k</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 32738 features across 2700 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (32738 features, 0 variable features)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="populate-a-somacollection">Populate a SOMACollection<a class="anchor" aria-label="anchor" href="#populate-a-somacollection"></a>
</h3>
<p>Our first step is to create a new <code>SOMACollection</code> at a
specific URI. The URI could be a local filepath, an S3 URI, or as
discussed below, a TileDB Cloud URI that automatically registers the
dataset on TileDB Cloud.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soco_uri</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"soco-pbmc3k"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">soco</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SOMACollection.html">SOMACollection</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>uri <span class="op">=</span> <span class="va">soco_uri</span><span class="op">)</span></span>
<span><span class="co">#&gt; No SOMACollection currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k'</span></span>
<span><span class="co">#&gt; Creating new SOMACollection at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k'</span></span>
<span><span class="co">#&gt; No TileDBGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/uns'</span></span>
<span><span class="co">#&gt; Creating new TileDBGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/uns'</span></span>
<span><span class="va">soco</span></span>
<span><span class="co">#&gt; &lt;SOMACollection&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k </span></span>
<span><span class="co">#&gt;   groups: uns</span></span></code></pre></div>
<p>Now we use tiledbsc to automatically ingest the various components of
the <code>pbmc3k</code> object into the new
<code>SOMACollection</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soco</span><span class="op">$</span><span class="fu">from_seurat</span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="co">#&gt; No SOMA currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA'</span></span>
<span><span class="co">#&gt; Creating new SOMA at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA'</span></span>
<span><span class="co">#&gt; No AnnotationDataframe found at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="co">#&gt; No AnnotationDataframe found at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/var'</span></span>
<span><span class="co">#&gt; No AssayMatrixGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X'</span></span>
<span><span class="co">#&gt; Creating new AssayMatrixGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X'</span></span>
<span><span class="co">#&gt; No AnnotationMatrixGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obsm'</span></span>
<span><span class="co">#&gt; Creating new AnnotationMatrixGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obsm'</span></span>
<span><span class="co">#&gt; No AnnotationMatrixGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/varm'</span></span>
<span><span class="co">#&gt; Creating new AnnotationMatrixGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/varm'</span></span>
<span><span class="co">#&gt; No AnnotationPairwiseMatrixGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obsp'</span></span>
<span><span class="co">#&gt; Creating new AnnotationPairwiseMatrixGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obsp'</span></span>
<span><span class="co">#&gt; No AnnotationPairwiseMatrixGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/varp'</span></span>
<span><span class="co">#&gt; Creating new AnnotationPairwiseMatrixGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/varp'</span></span>
<span><span class="co">#&gt; No TileDBGroup currently exists at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/uns'</span></span>
<span><span class="co">#&gt; Creating new TileDBGroup at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/uns'</span></span>
<span><span class="co">#&gt; Creating new AnnotationDataframe array with index [obs_id] at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="co">#&gt; Ingesting AnnotationDataframe data into: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs</span></span>
<span><span class="co">#&gt; Creating new AnnotationDataframe array with index [var_id] at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/var'</span></span>
<span><span class="co">#&gt; Ingesting AnnotationDataframe data into: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/var</span></span>
<span><span class="co">#&gt; Skipping ingestion of 'data' because it is identical to 'counts'</span></span>
<span><span class="co">#&gt; No AssayMatrix found at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts'</span></span>
<span><span class="co">#&gt; Creating new AssayMatrix array with index [var_id,obs_id] at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts'</span></span>
<span><span class="co">#&gt; Ingesting AssayMatrix data into: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts</span></span>
<span><span class="co">#&gt; Finished converting Seurat Assay with key [rna_] to SOMA</span></span>
<span><span class="co">#&gt; Finished converting Seurat object to SOMACollection</span></span></code></pre></div>
<p>Printing the <code>soco</code> object shows it comprises 2 members:
<code>RNA</code>, which contains the RNA assay data, and
<code>uns</code>, which is a special group for miscellaneous (i.e.,
unstructured) data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soco</span></span>
<span><span class="co">#&gt; &lt;SOMACollection&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k </span></span>
<span><span class="co">#&gt;   groups: RNA, uns</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="access-soma-components">Access SOMA Components<a class="anchor" aria-label="anchor" href="#access-soma-components"></a>
</h3>
<p>As this dataset is unimodal, we’ll extract the SOMA containing the
RNA data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span> <span class="op">&lt;-</span> <span class="va">soco</span><span class="op">$</span><span class="fu">get_member</span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="va">soma</span></span>
<span><span class="co">#&gt; &lt;SOMA&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA </span></span>
<span><span class="co">#&gt;   arrays: obs, var </span></span>
<span><span class="co">#&gt;   groups: obsm, obsp, uns, varm, varp, X</span></span></code></pre></div>
<p>Access the [<code>AnnotationDataframe</code>] representing the
<code>obs</code> array containing cell-level metadata stored in
<code>obs</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">obs</span></span>
<span><span class="co">#&gt; &lt;AnnotationDataframe&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs </span></span>
<span><span class="co">#&gt;   dimensions: obs_id </span></span>
<span><span class="co">#&gt;   attributes: orig.ident, nCount_RNA, nFeature_RNA</span></span></code></pre></div>
<p>Read it into memory as a <code>data.frame</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obs</span> <span class="op">&lt;-</span> <span class="va">soma</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="fu">to_dataframe</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">obs</span><span class="op">)</span></span>
<span><span class="co">#&gt;   orig.ident          nCount_RNA     nFeature_RNA   </span></span>
<span><span class="co">#&gt;  Length:2700        Min.   :  548   Min.   : 212.0  </span></span>
<span><span class="co">#&gt;  Class :character   1st Qu.: 1758   1st Qu.: 690.0  </span></span>
<span><span class="co">#&gt;  Mode  :character   Median : 2197   Median : 817.0  </span></span>
<span><span class="co">#&gt;                     Mean   : 2367   Mean   : 847.0  </span></span>
<span><span class="co">#&gt;                     3rd Qu.: 2763   3rd Qu.: 953.2  </span></span>
<span><span class="co">#&gt;                     Max.   :15844   Max.   :3422.0</span></span></code></pre></div>
<p>Access the [<code>AssayMatrix</code>] representing the <code>X</code>
array containing the raw RNA counts.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">X</span><span class="op">$</span><span class="va">members</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="co">#&gt; &lt;AssayMatrix&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts </span></span>
<span><span class="co">#&gt;   dimensions: var_id, obs_id </span></span>
<span><span class="co">#&gt;   attributes: value</span></span></code></pre></div>
<p>Read it into memory as a
[<code>dgTMatrix</code>][<code>Matrix::TsparseMatrix-class</code>].</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="va">soma</span><span class="op">$</span><span class="va">X</span><span class="op">$</span><span class="va">members</span><span class="op">$</span><span class="va">counts</span><span class="op">$</span><span class="fu">to_matrix</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading AssayMatrix into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 16634  2700</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dimension-slicing-and-attribute-filtering">Dimension Slicing and Attribute Filtering<a class="anchor" aria-label="anchor" href="#dimension-slicing-and-attribute-filtering"></a>
</h3>
<p><code>SOMACollection</code> and <code>SOMA</code> objects can be
sliced by a combination of obs and var identifiers.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">set_query</span><span class="op">(</span>obs_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This slice is automatically applied to the entire SOMA, so any
component read into memory will now only contain the filtered
results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="fu">to_dataframe</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1 SeuratProject       2421          781</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1 SeuratProject       4903         1352</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1 SeuratProject       3149         1131</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1 SeuratProject       2639          960</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1 SeuratProject        981          522</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1 SeuratProject       2164          782</span></span>
<span><span class="co">#&gt; AAACGCTGACCAGT-1 SeuratProject       2176          783</span></span>
<span><span class="co">#&gt; AAACGCTGGTTCTT-1 SeuratProject       2260          790</span></span>
<span><span class="co">#&gt; AAACGCTGTAGCCA-1 SeuratProject       1276          533</span></span>
<span><span class="co">#&gt; AAACGCTGTTTCTG-1 SeuratProject       1103          550</span></span></code></pre></div>
<p>To remove the filter simply reset the query.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">reset_query</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">soma</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="fu">to_dataframe</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="co">#&gt;   orig.ident          nCount_RNA     nFeature_RNA   </span></span>
<span><span class="co">#&gt;  Length:2700        Min.   :  548   Min.   : 212.0  </span></span>
<span><span class="co">#&gt;  Class :character   1st Qu.: 1758   1st Qu.: 690.0  </span></span>
<span><span class="co">#&gt;  Mode  :character   Median : 2197   Median : 817.0  </span></span>
<span><span class="co">#&gt;                     Mean   : 2367   Mean   : 847.0  </span></span>
<span><span class="co">#&gt;                     3rd Qu.: 2763   3rd Qu.: 953.2  </span></span>
<span><span class="co">#&gt;                     Max.   :15844   Max.   :3422.0</span></span></code></pre></div>
<p>In addition to dimension slicing we can also apply filters using the
cell- and feature-level metadata stored in the <code>obs</code> and
<code>var</code> arrays.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">set_query</span><span class="op">(</span></span>
<span>    obs_attr_filter <span class="op">=</span> <span class="va">seurat_annotations</span> <span class="op">==</span> <span class="st">"B"</span> <span class="op">&amp;&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Querying obs with attribute filter</span></span>
<span><span class="co">#&gt; Error: No attibute 'seurat_annotations' present.</span></span></code></pre></div>
<p>We can load the SOMA slice into memory as a Seurat
<code>Assay</code>…</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">to_seurat_assay</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading AssayMatrix into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts'</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/var'</span></span>
<span><span class="co">#&gt; Assay data with 16634 features for 2700 cells</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;  ENSG00000000003, ENSG00000000419, ENSG00000000457, ENSG00000000460,</span></span>
<span><span class="co">#&gt; ENSG00000000938, ENSG00000000971, ENSG00000001036, ENSG00000001084,</span></span>
<span><span class="co">#&gt; ENSG00000001167, ENSG00000001460</span></span></code></pre></div>
<p>or as a Bioconductor [<code>SummarizedExperiment</code>].</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"SummarizedExperiment"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">soma</span><span class="op">$</span><span class="fu">to_summarized_experiment</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Reading AssayMatrix into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/counts'</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obs'</span></span>
<span><span class="co">#&gt; Reading AnnotationDataframe into memory from '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/var'</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 16634 2700 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(16634): ENSG00000000003 ENSG00000000419 ... ENSG00000273486</span></span>
<span><span class="co">#&gt;   ENSG00000273489</span></span>
<span><span class="co">#&gt; rowData names(1): gene_name</span></span>
<span><span class="co">#&gt; colnames(2700): ATAGATACCATGGT-1 AAACATTGATCAGC-1 ... ACTTGTACCTGTCC-1</span></span>
<span><span class="co">#&gt;   GCAACCCTCCTCGT-1</span></span>
<span><span class="co">#&gt; colData names(3): orig.ident nCount_RNA nFeature_RNA</span></span></code></pre></div>
<p>See <code><a href="../articles/filtering.html">vignette("filtering", package = "tiledbsc")</a></code> for more
details.</p>
</div>
<div class="section level3">
<h3 id="adding-analysis-results-to-a-soma">Adding Analysis Results to a SOMA<a class="anchor" aria-label="anchor" href="#adding-analysis-results-to-a-soma"></a>
</h3>
<p>The original <code>pbmc3k</code> object contains only the raw counts
data. Here we’ll use <code>Seurat</code> to generate new results and add
them to the <code>soma</code>.</p>
<div class="section level4">
<h4 id="add-layer-for-normalized-counts">Add Layer for Normalized Counts<a class="anchor" aria-label="anchor" href="#add-layer-for-normalized-counts"></a>
</h4>
<p>First we normalize the raw counts, which creates a new matrix stored
in the <code>Assay</code> object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">NormalizeData</span><span class="op">(</span><span class="va">pbmc3k</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span></code></pre></div>
<p>Currently the RNA SOMA’s <code>X</code> group contains only a single
layer, <code>counts</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">X</span></span>
<span><span class="co">#&gt; &lt;AssayMatrixGroup&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X </span></span>
<span><span class="co">#&gt;   arrays: counts</span></span></code></pre></div>
<p>Add a new <code>X</code> layer containing log-normalized counts
stored in the Seurat <code>Assay</code>’s <code>data</code> slot.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">from_seurat_assay</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  layers <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  var <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; No AssayMatrix found at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/data'</span></span>
<span><span class="co">#&gt; Creating new AssayMatrix array with index [var_id,obs_id] at '/var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/data'</span></span>
<span><span class="co">#&gt; Ingesting AssayMatrix data into: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X/data</span></span>
<span><span class="co">#&gt; Finished converting Seurat Assay with key [rna_] to SOMA</span></span></code></pre></div>
<p>Now we can see the SOMA contains 2 <code>X</code> layers:
<code>counts</code> and <code>data</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">X</span></span>
<span><span class="co">#&gt; &lt;AssayMatrixGroup&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X </span></span>
<span><span class="co">#&gt;   arrays: counts, data</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-layer-for-scaled-counts">Add Layer for Scaled counts<a class="anchor" aria-label="anchor" href="#add-layer-for-scaled-counts"></a>
</h4>
<p>First, we identify highly variable subset of genes to scale.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span></span>
<span>  <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">pbmc3k</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Then generate the scaled data.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">ScaleData</span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="va">var_genes</span><span class="op">)</span></span></code></pre></div>
<p>And ingest the scaled data into a new <code>X</code> layer called
<code>"scale.data"</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">from_seurat_assay</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    layers <span class="op">=</span> <span class="st">"scale.data"</span>,</span>
<span>    var <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Again, verify the new layer was added.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">X</span></span>
<span><span class="co">#&gt; &lt;AssayMatrixGroup&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/X </span></span>
<span><span class="co">#&gt;   arrays: counts, data</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-dimensional-reduction-results">Add Dimensional Reduction Results<a class="anchor" aria-label="anchor" href="#add-dimensional-reduction-results"></a>
</h4>
<p>We can use a similar workflow for dimensional-reduction results,
which are stored in either the <code>obsm</code> or <code>varm</code>
slots of a SOMA.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dimreduc_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Reductions</a></span><span class="op">(</span></span>
<span>  <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">RunPCA</span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="va">var_genes</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"pca"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Add the Seurat <code>DimReduc</code> object to the SOMA.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="fu">add_seurat_dimreduction</span><span class="op">(</span></span>
<span>    <span class="va">dimreduc_pca</span>,</span>
<span>    technique <span class="op">=</span> <span class="st">"pca"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The SOMA now contains 2 new arrays, one within the <code>obsm</code>
group storing the cell embeddings</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">obsm</span></span>
<span><span class="co">#&gt; &lt;AnnotationMatrixGroup&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/obsm</span></span></code></pre></div>
<p>and a second within the <code>varm</code> group storing the feature
loadings.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soma</span><span class="op">$</span><span class="va">varm</span></span>
<span><span class="co">#&gt; &lt;AnnotationMatrixGroup&gt;</span></span>
<span><span class="co">#&gt;   uri: /var/folders/mb/hbl3p_1d4kv4_plbcqb6wd8w0000gn/T//RtmpOA4uOo/soco-pbmc3k/soma_RNA/varm</span></span></code></pre></div>
<p>Either of these can be accessed and queried directly, which is often
useful for visualization applications.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obsm_pca</span> <span class="op">&lt;-</span> <span class="va">soma</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="fu">get_member</span><span class="op">(</span><span class="st">"dimreduction_pca"</span><span class="op">)</span></span>
<span><span class="va">obsm_pca</span></span></code></pre></div>
<p>Here we’ll load the cell embeddings as a <code>matrix</code> and
visualize the first 5 PCs.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_pcs</span> <span class="op">&lt;-</span> <span class="va">obsm_pca</span><span class="op">$</span><span class="fu">to_matrix</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html" class="external-link">pairs</a></span><span class="op">(</span><span class="va">cell_pcs</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="populate-a-somacollection-in-tiledb-cloud">Populate a SOMACollection in TileDB Cloud<a class="anchor" aria-label="anchor" href="#populate-a-somacollection-in-tiledb-cloud"></a>
</h3>
<p>You can ingest a Seurat object into an S3 <code>SOMACollection</code>
that’s automatically registered with TileDB Cloud using a special URI
composed of:</p>
<ul>
<li>your TileDB namespace</li>
<li>an S3 bucket where the dataset will be created</li>
</ul>
<p>and running the same initialization/ingestion steps we performed
earlier:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soco</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SOMACollection.html">SOMACollection</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="st">"tiledb://&lt;namespace&gt;/s3://&lt;bucket&gt;/pbmc3k-soco"</span><span class="op">)</span></span>
<span><span class="va">soco</span><span class="op">$</span><span class="fu">from_seurat</span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span></code></pre></div>
<p>Note this assumes you:</p>
<ul>
<li>have a TileDB Cloud account</li>
<li>registered <a href="https://cloud.tiledb.com/user-settings/credentials" class="external-link">cloud
credentials</a> for AWS</li>
<li>provided a <a href="https://cloud.tiledb.com/user-settings/storage-paths" class="external-link">default S3
storage path</a>
</li>
<li>are <a href="https://tiledb-inc.github.io/TileDB-Cloud-R/articles/Login.html" class="external-link">logged
into TileDB Cloud</a> in your environment</li>
</ul>
<p>You can then view the new SOMACollection directly on TileDB Cloud
where you can inspect the various dataset components, share securely,
view activity logs, and more. For example, here’s the TileDB Cloud view
of the <code>pbmc3k</code> dataset we just ingested: <a href="https://cloud.tiledb.com/groups/aaronwolen/68ffab71-b090-42c9-8d8b-977d00979257/overview" class="external-link"><code>aaronwolen/pbmc3k-soco</code></a>.</p>
<p>You can also securely access the dataset directly via R (or Python)
using its TileDB Cloud URI,
<code>tiledb://aaronwolen/pbmc3k-soco</code>.</p>
</div>
<div class="section level2">
<h2 id="session">Session<a class="anchor" aria-label="anchor" href="#session"></a>
</h2>
<details><summary>
Session Info
</summary><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] RcppSpdlog_0.0.12  SeuratObject_4.1.3 sp_1.6-0           tiledbsc_0.1.5    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Rcpp_1.0.10                 XVector_0.38.0             </span></span>
<span><span class="co">#&gt;  [3] GenomeInfoDb_1.34.9         compiler_4.2.2             </span></span>
<span><span class="co">#&gt;  [5] progressr_0.13.0            zlibbioc_1.44.0            </span></span>
<span><span class="co">#&gt;  [7] bitops_1.0-7                MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt;  [9] tools_4.2.2                 digest_0.6.31              </span></span>
<span><span class="co">#&gt; [11] bit_4.0.5                   evaluate_0.20              </span></span>
<span><span class="co">#&gt; [13] lattice_0.20-45             nanotime_0.3.7             </span></span>
<span><span class="co">#&gt; [15] rlang_1.0.6                 Matrix_1.5-3               </span></span>
<span><span class="co">#&gt; [17] DelayedArray_0.24.0         cli_3.6.0                  </span></span>
<span><span class="co">#&gt; [19] RcppCCTZ_0.2.12             spdl_0.0.4                 </span></span>
<span><span class="co">#&gt; [21] parallel_4.2.2              xfun_0.37                  </span></span>
<span><span class="co">#&gt; [23] GenomeInfoDbData_1.2.9      knitr_1.42                 </span></span>
<span><span class="co">#&gt; [25] IRanges_2.32.0              S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt; [27] fs_1.6.1                    vctrs_0.5.2                </span></span>
<span><span class="co">#&gt; [29] globals_0.16.2              stats4_4.2.2               </span></span>
<span><span class="co">#&gt; [31] triebeard_0.3.0             bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [33] grid_4.2.2                  Biobase_2.58.0             </span></span>
<span><span class="co">#&gt; [35] glue_1.6.2                  data.table_1.14.8          </span></span>
<span><span class="co">#&gt; [37] listenv_0.9.0               R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [39] future.apply_1.10.0         parallelly_1.34.0          </span></span>
<span><span class="co">#&gt; [41] tiledb_0.18.0.1             GenomicRanges_1.50.2       </span></span>
<span><span class="co">#&gt; [43] matrixStats_0.63.0          urltools_1.7.3             </span></span>
<span><span class="co">#&gt; [45] codetools_0.2-18            BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt; [47] SummarizedExperiment_1.28.0 future_1.32.0              </span></span>
<span><span class="co">#&gt; [49] RCurl_1.98-1.10             zoo_1.8-11</span></span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Wolen, Dirk Eddelbuettel, John Kerl, TileDB, Inc..</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
